packages/runtime/container-runtime/src/containerRuntime.ts

```diff
@@ -3479,10 +3479,20 @@ export class ContainerRuntime
 		return result;
 	}
 
+	private stageControls: StageControls | undefined;
```

Comment:
Not exactly related to this PR but made me think about it since you're retaining the reference now - should we do something with these on dispose()?  E.g. call discardChanges()?

Reply:
we probably should do something. we do some interesting things with blobs and closing esp when gathering pending state where we enter a shutdown mode, as ideally the app decides what to do, i would probably block the pending state on the app deciding, and for normal close discard should be fine